tape/src/main/java/com/squareup/tape/QueueFile.java

```diff
@@ -523,25 +523,34 @@ public synchronized void remove() throws IOException {
     if (elementCount == 1) {
       clear();
     } else {
-      // assert elementCount > 1
       int firstTotalLength = Element.HEADER_LENGTH + first.length;
 
-      ringErase(first.position, firstTotalLength);
-
+      // Read the next element
       int newFirstPosition = wrapPosition(first.position + firstTotalLength);
       ringRead(newFirstPosition, buffer, 0, Element.HEADER_LENGTH);
-      int length = readInt(buffer, 0);
+      int newFirstLength = readInt(buffer, 0);
+
+      // Commit the headers
       writeHeader(fileLength, elementCount - 1, newFirstPosition, last.position);
+
+      Element removed = first;
       elementCount--;
-      first = new Element(newFirstPosition, length);
+      first = new Element(newFirstPosition, newFirstLength);
```

Comment:
Done eagerly so that if `ringErase`ing the old first element throws an Exception, the caller can still catch it and continue using the file in a valid state.